### ISSUE TITLE
YouTube: HTTP Error 401: Unauthorized

### ISSUE THREAD

**Issue by sixfootdad** [coletdjnz/yt-dlp-youtube-oauth2#38]
Hello - I'm getting HTTP 401 when attempting to login using oauth2. Has YT blocked the `_CLIENT_ID` or `_CLIENT_SECRET`?

I'm using the latest release of this plugin and yesterday's yt-dlp nightly.
```
[debug] Encodings: locale utf-8, fs utf-8, pref utf-8, out utf-8 (No ANSI), error utf-8 (No ANSI), screen utf-8 (No ANSI)
[debug] yt-dlp version nightly@2024.09.02.232855 from yt-dlp/yt-dlp-nightly-builds [e8e6a982a] API
[debug] params: {'outtmpl': '/tmp/PgbkQYXfbBI.mp4', 'cachedir': '/tmp', 'ffmpeg_location': '/opt/python/lib/python3.12/site-packages/ffmpeg-master-latest-linux64-gpl/bin', 'verbose': True, 'username': 'oauth2', 'password': '', 'ignoreerrors': True, 'merge_output_format': 'mp4', 'compat_opts': set(), 'http_headers': {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.50 Safari/537.36', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8', 'Accept-Language': 'en-us,en;q=0.5', 'Sec-Fetch-Mode': 'navigate'}}
[debug] Python 3.12.4 (CPython x86_64 64bit) - Linux-5.10.223-232.872.amzn2.x86_64-x86_64-with-glibc2.34 (OpenSSL 3.0.8 7 Feb 2023, glibc 2.34)
[debug] exe versions: ffmpeg N-116556-gcda5f5c5ed-20240809 (setts)
[debug] Optional libraries: Cryptodome-3.20.0, certifi-2024.07.04, mutagen-1.47.0, requests-2.32.3, sqlite3-3.40.0, urllib3-2.2.2, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Extractor Plugins: oauth2 (YoutubeClipIE), oauth2 (YoutubeConsentRedirectIE), oauth2 (YoutubeFavouritesIE), oauth2 (YoutubeIE), oauth2 (YoutubeMusicSearchURLIE), oauth2 (YoutubeNotificationsIE), oauth2 (YoutubeSearchDateIE), oauth2 (YoutubeSearchIE), oauth2 (YoutubeSearchURLIE), oauth2 (YoutubeTabIE)
[debug] Plugin directories: ['/opt/python/lib/python3.12/site-packages/yt_dlp_plugins']
[debug] Loaded 1832 extractors
[debug] YouTube OAuth2 plugin version 2024.08.31.1
[youtube+oauth2] oauth2: Initializing OAuth2 Authorization Flow
ERROR: [youtube+oauth2] PgbkQYXfbBI: Unable to download JSON metadata: HTTP Error 401: Unauthorized (caused by <HTTPError 401: Unauthorized>)
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 737, in extract
self.initialize()
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 643, in initialize
self._perform_login(username, password)
File "/opt/python/lib/python3.12/site-packages/yt_dlp_plugins/extractor/youtubeoauth.py", line 192, in _perform_login
self.initialize_oauth()
File "/opt/python/lib/python3.12/site-packages/yt_dlp_plugins/extractor/youtubeoauth.py", line 72, in initialize_oauth
token_data = self.authorize()
^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp_plugins/extractor/youtubeoauth.py", line 130, in authorize
code_response = self._download_json(
^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 1139, in download_content
res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 1099, in download_handle
res = self._download_webpage_handle(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 960, in _download_webpage_handle
urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data,
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 909, in _request_webpage
raise ExtractorError(errmsg, cause=err)
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 896, in _request_webpage
return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query, extensions))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/YoutubeDL.py", line 4165, in urlopen
return self._request_director.send(req)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/networking/common.py", line 117, in send
response = handler.send(request)
^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/networking/_helper.py", line 208, in wrapper
return func(self, *args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/networking/common.py", line 340, in send
return self._send(request)
^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/networking/_requests.py", line 365, in _send
raise HTTPError(res, redirect_loop=max_redirects_exceeded)
yt_dlp.networking.exceptions.HTTPError: HTTP Error 401: Unauthorized
```

Here's the debug when I try to download without oauth2. As expected, YT asks for a sign-in.
```
[debug] Encodings: locale utf-8, fs utf-8, pref utf-8, out utf-8 (No ANSI), error utf-8 (No ANSI), screen utf-8 (No ANSI)
[debug] yt-dlp version nightly@2024.09.02.232855 from yt-dlp/yt-dlp-nightly-builds [e8e6a982a] API
[debug] params: {'outtmpl': '/tmp/PgbkQYXfbBI.mp4', 'cachedir': '/tmp', 'ffmpeg_location': '/opt/python/lib/python3.12/site-packages/ffmpeg-master-latest-linux64-gpl/bin', 'verbose': True, 'ignoreerrors': True, 'merge_output_format': 'mp4', 'compat_opts': set(), 'http_headers': {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.164 Safari/537.36', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8', 'Accept-Language': 'en-us,en;q=0.5', 'Sec-Fetch-Mode': 'navigate'}}
[debug] Python 3.12.4 (CPython x86_64 64bit) - Linux-5.10.223-232.872.amzn2.x86_64-x86_64-with-glibc2.34 (OpenSSL 3.0.8 7 Feb 2023, glibc 2.34)
[debug] exe versions: ffmpeg N-116556-gcda5f5c5ed-20240809 (setts)
[debug] Optional libraries: Cryptodome-3.20.0, certifi-2024.07.04, mutagen-1.47.0, requests-2.32.3, sqlite3-3.40.0, urllib3-2.2.2, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Extractor Plugins: oauth2 (YoutubeClipIE), oauth2 (YoutubeConsentRedirectIE), oauth2 (YoutubeFavouritesIE), oauth2 (YoutubeIE), oauth2 (YoutubeMusicSearchURLIE), oauth2 (YoutubeNotificationsIE), oauth2 (YoutubeSearchDateIE), oauth2 (YoutubeSearchIE), oauth2 (YoutubeSearchURLIE), oauth2 (YoutubeTabIE)
[debug] Plugin directories: ['/opt/python/lib/python3.12/site-packages/yt_dlp_plugins']
[debug] Loaded 1832 extractors
[debug] YouTube OAuth2 plugin version 2024.08.31.1
[youtube+oauth2] Extracting URL: https://www.youtube.com/watch?v=PgbkQYXfbBI
[youtube+oauth2] PgbkQYXfbBI: Downloading webpage
[youtube+oauth2] PgbkQYXfbBI: Downloading ios player API JSON
[youtube+oauth2] PgbkQYXfbBI: Downloading web creator player API JSON
ERROR: [youtube+oauth2] PgbkQYXfbBI: Sign in to confirm you’re not a bot. This helps protect our community. Learn more
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 740, in extract
ie_result = self._real_extract(url)
^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/youtube.py", line 4366, in _real_extract
self.raise_no_formats(reason, expected=True)
File "/opt/python/lib/python3.12/site-packages/yt_dlp/extractor/common.py", line 1263, in raise_no_formats
raise ExtractorError(msg, expected=expected, video_id=video_id)
```

**Comment by coletdjnz** [comment]
Can you try logging into youtube TV in your browser and see if it still works there (youtube.com/tv, with a user agent switcher extension)?

Might also be worth checking if using an IP that is not blocked works?

**Comment by sixfootdad** [comment]
I went to youtube.com/tv in my desktop browser (Firefox v130.0) with the user-agent switched to `Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36` which just redirects to https://www.youtube.com/?app=desktop

In my case, yt-dlp is actually running on an AWS Lambda function, which makes it difficult to debug in general. Of course, yt-dlp downloads videos just fine from my desktop without forcing me to login.

**Comment by coletdjnz** [comment]
> I went to youtube.com/tv in my desktop browser (Firefox v130.0) with the user-agent switched to Mozilla/5.0 (Windows NT 10.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36 which just redirects to https://www.youtube.com/?app=desktop

I was meaning use a TV user agent - Youtube redirects to desktop otherwise.

I'm unable to reproduce - it works fine for me. Perhaps youtube is blocking oauth on that IP? You'll need to VPN in and test on a web browser (with a TV UA) to confirm.

**Comment by sixfootdad** [comment]
So I've got good news and bad news. The good news is that I got it working. I fired up an Amazon Linux instance on EC2, ssh'd in, installed yt-dlp & this oauth2 plugin, successfully authenticated using the YouTubeTV client id, and then copied the details out of `/root/.cache/yt-dlp/youtube-oauth2/token_data.json` to reuse on AWS Lambda. (I've even stored the access & refresh tokens as encrypted environment variables and then write out a new `token_data.json` to the cachedir specified in my yt-dlp config.) This all works splendidly, and I'm able to refresh the token as needed.

The bad news? I have no idea why I was getting the HTTP 401 on Lambda as it's effectively ephemeral compute, and there's no way to login to a Lambda instance while it's running.

tl;dr - I got it working, but don't know why it wasn't. I'm closing the issue. :)
